server/api/src/rsDataObjUnlink.cpp

```diff
@@ -547,125 +564,95 @@ int dataObjUnlinkS(rsComm_t* rsComm,
+    // in this implementation. Just return success.
+    if (dataObjInfo->specColl) {
+        return 0;
     }
 
-    if ( dataObjUnlinkInp->oprType != UNREG_OPR ) {
-        // Set the in_pdmo flag
-        char* in_pdmo = getValByKey( &dataObjUnlinkInp->condInput, IN_PDMO_KW );
-        if ( in_pdmo != NULL ) {
-            rstrcpy( dataObjInfo->in_pdmo, in_pdmo, MAX_NAME_LEN );
-        }
-        else {
-            dataObjInfo->in_pdmo[0] = '\0';
-        }
+    unregDataObj_t unreg_inp{};
+    unreg_inp.dataObjInfo = info.get();
+    unreg_inp.condInput = &dataObjUnlinkInp->condInput;
+    if (const auto ec = rsUnregDataObj(rsComm, &unreg_inp); ec < 0) {
+        irods::log(LOG_NOTICE, fmt::format(
+            "[{}:{}] - rsUnregDataObj failed "
+            "[error_code=[{}], path=[{}], hierarchy=[{}]",
+            __FUNCTION__, __LINE__, ec, info.logical_path(), info.hierarchy()));
 
-        status = l3Unlink( rsComm, dataObjInfo );
-        if ( status < 0 ) {
-            int myError = getErrno( status );
-            rodsLog( LOG_NOTICE,
-                     "dataObjUnlinkS: l3Unlink error for %s. status = %d",
-                     dataObjUnlinkInp->objPath, status );
-            /* allow ENOENT to go on and unregister */
-            if ( myError != ENOENT && myError != EACCES ) {
-                char orphanPath[MAX_NAME_LEN];
-                int status1 = 0;
-                rodsLog( LOG_NOTICE,
-                         "dataObjUnlinkS: orphan file %s", dataObjInfo->filePath );
-                while ( 1 ) {
-                    if ( isOrphanPath( dataObjUnlinkInp->objPath ) ==
-                            NOT_ORPHAN_PATH ) {
-                        /* don't rename orphan path */
-                        status1 = rsMkOrphanPath( rsComm, dataObjInfo->objPath,
-                                                  orphanPath );
-                        if ( status1 < 0 ) {
-                            break;
-                        }
-                        /* reg the orphan path */
-                        rstrcpy( dataObjInfo->objPath, orphanPath, MAX_NAME_LEN );
-                    }
-                    status1 = svrRegDataObj( rsComm, dataObjInfo );
-                    if ( status1 == CAT_NAME_EXISTS_AS_DATAOBJ ||
-                            status1 == CATALOG_ALREADY_HAS_ITEM_BY_THAT_NAME ) {
-                        continue;
-                    }
-                    else if ( status1 < 0 ) {
-                        rodsLogError( LOG_ERROR, status1,
-                                      "dataObjUnlinkS: svrRegDataObj of orphan %s error",
-                                      dataObjInfo->objPath );
-                    }
-                    break;
-                }
-                return status;
-            }
-            else {
-                status = 0;
-            }
-        }
+        // TODO: Should the replica be marked stale if the unregister itself fails?
```

Comment:
if the data on disk and the entry in the catalog are known to be good... we should leave it good.   for *anything* other case, we can't vouch for it, so yes, stale it.

Reply:
I think I will revisit this on the pass to prevent unlinks on locked data objects.